### ISSUE TITLE
Remove 'extents' VLR as a required COPC entity

### ISSUE THREAD

**Issue by hobu** [copcio/copcio.github.io#47]
`extents` by itself isn't particularly useful in practice. Knowing the min/max only helps you for a few dimensions, and if you're a renderer, you want much more than just that. 

I think we should drop `extents` and put forward a definition for a `copc` `stats` VLR that is more fully featured but OPTIONAL. Importantly, the `stats` VLR could be something that could be added as an EVLR long after the file is created instead of during the COPC file creation process. 

Principles of the `stats` VLR are pretty much as @connormanning suggested in #19, but we should do something specified as JSON, with a corresponding [JSON Schema](https://json-schema.org/) that allows implementations freedom to augment with auxiliary information if they desire.

**Comment by hobu** [comment]
Here' is a straw man to kick around. I chose this layout for a few reasons:

* jq query friendly
* easy to extend with custom attributes and supporting information if desired

I'm not wed to it at all, however. I'm happy to use some other reasonable alternative. If there is already something in practical use, that would be even better.

```
[
    {
        "name": "X",
        "value": {
            "histogram": {
                "bins": [
                    1434.9832443226683,
                    1495.9322800856594,
                    1556.8813158486503,
                    1617.8303516116414,
                    1678.7793873746325,
                    1739.7284231376234,
                    1800.6774589006145,
                    1861.6264946636056,
                    1922.5755304265967,
                    1983.5245661895876,
                    2044.4736019525787
                ],
                "counts": [
                    3,
                    2,
                    0,
                    1,
                    1,
                    2,
                    2,
                    2,
                    3,
                    3
                ]
            },
            "statistics": [
                {
                    "name": "minimum",
                    "value": 1434.9832443226683
                },
                {
                    "name": "maximum",
                    "value": 2044.4736019525787
                },
                {
                    "name": "mean",
                    "value": 1770.5773071877832
                },
                {
                    "name": "stddev",
                    "value": 201.46056835843007
                },
                {
                    "name": "median",
                    "value": 1848.41355112089
                }
            ],
            "sample": [
                1457.667401303338,
                1984.5820959500481,
                1874.5564339385865,
                1914.269878323662,
                1850.3610272931605,
                2029.4279446355968,
                1685.0915939208126,
                1848.41355112089,
                1644.8120411335271,
                1685.0915939208126,
                1956.1265270178415,
                1956.1265270178415,
                1984.5820959500481,
                1751.766293544963,
                1984.5820959500481,
                2044.4736019525787,
                1914.269878323662,
                1457.667401303338,
                1644.8120411335271,
                1519.0500565263349
            ]
        }
    },
    {
        "name": "Y",
        "value": {
            "histogram": {
                "bins": [
                    1154.9878678258765,
                    1501.1357412814673,
                    1847.2836147370579,
                    2193.4314881926484,
                    2539.5793616482392,
                    2885.72723510383,
                    3231.8751085594204,
                    3578.022982015011,
                    3924.170855470602,
                    4270.318728926193,
                    4616.466602381784
                ],
                "counts": [
                    3,
                    2,
                    1,
                    2,
                    2,
                    1,
                    0,
                    1,
                    2,
                    5
                ]
            },
            "statistics": [
                {
                    "name": "minimum",
                    "value": 1154.9878678258765
                },
                {
                    "name": "maximum",
                    "value": 4616.466602381784
                },
                {
                    "name": "mean",
                    "value": 2980.978433100374
                },
                {
                    "name": "stddev",
                    "value": 1220.871288116911
                },
                {
                    "name": "median",
                    "value": 2879.8521169671812
                }
            ],
            "sample": [
                4320.288865237972,
                1156.851576076897,
                2879.8521169671812,
                4043.861992863049,
                3231.577715939853,
                4320.288865237972,
                1932.397675515459,
                3593.8626116729865,
                4043.861992863049,
                1781.4825383708953,
                3231.577715939853,
                4320.288865237972,
                1156.851576076897,
                1781.4825383708953,
                1781.4825383708953,
                4616.466602381784,
                1781.4825383708953,
                2544.6956591279,
                2509.738044923516,
                4496.128428793121
            ]
        }
    },
    {
        "name": "Intensity",
        "value": {
            "histogram": {
                "bins": [
                    2814.170506233245,
                    2967.683636336252,
                    3121.196766439259,
                    3274.7098965422665,
                    3428.2230266452734,
                    3581.736156748281,
                    3735.249286851288,
                    3888.762416954295,
                    4042.2755470573024,
                    4195.78867716031,
                    4349.301807263317
                ],
                "counts": [
                    5,
                    2,
                    1,
                    1,
                    2,
                    1,
                    2,
                    1,
                    2,
                    2
                ]
            },
            "statistics": [
                {
                    "name": "minimum",
                    "value": 2814.170506233245
                },
                {
                    "name": "maximum",
                    "value": 4349.301807263317
                },
                {
                    "name": "mean",
                    "value": 3480.9312029085313
                },
                {
                    "name": "stddev",
                    "value": 517.3360960608679
                },
                {
                    "name": "median",
                    "value": 3475.429125278882
                }
            ],
            "sample": [
                3067.8700953530824,
                2864.479903763585,
                3797.6580592370988,
                3574.5873925363558,
                3806.223021621833,
                3475.429125278882,
                4349.301807263317,
                3203.169293964371,
                3279.464087360171,
                3806.223021621833,
                4019.2716587119053,
                2868.9049917524258,
                2868.9049917524258,
                2864.479903763585,
                3797.6580592370988,
                3067.8700953530824,
                4019.2716587119053,
                4122.076014844328,
                3797.6580592370988,
                3806.223021621833
            ]
        }
    },
    {
        "name": "PointSourceId",
        "value": {
            "histogram": {
                "bins": [
                    2112.3863406278724,
                    2319.9795289442063,
                    2527.5727172605402,
                    2735.165905576874,
                    2942.759093893208,
                    3150.352282209542,
                    3357.945470525876,
                    3565.5386588422098,
                    3773.1318471585437,
                    3980.7250354748776,
                    4188.3182237912115
                ],
                "counts": [
                    1,
                    1,
                    2,
                    3,
                    4,
                    3,
                    2,
                    0,
                    1,
                    2
                ]
            },
            "statistics": [
                {
                    "name": "minimum",
                    "value": 2112.3863406278724
                },
                {
                    "name": "maximum",
                    "value": 4188.3182237912115
                },
                {
                    "name": "mean",
                    "value": 3114.744921147889
                },
                {
                    "name": "stddev",
                    "value": 524.4270346033472
                },
                {
                    "name": "median",
                    "value": 3076.9384581610707
                }
            ],
            "sample": [
                3296.988842108107,
                2958.135720708723,
                3076.9384581610707,
                2958.135720708723,
                2756.6886716100976,
                2537.1269443602605,
                2566.359038903229,
                2112.3863406278724,
                4188.3182237912115,
                2958.135720708723,
                4188.3182237912115,
                3102.7773056491233,
                4139.483077468094,
                3263.4040879221357,
                3406.031078754512,
                2958.135720708723,
                2895.013994242619,
                3102.7773056491233,
                2566.359038903229,
                2756.6886716100976
            ]
        }
    },
    {
        "name": "GPSTime",
        "value": {
            "histogram": {
                "bins": [
                    2483.69748133072,
                    2677.00625026821,
                    2870.315019205699,
                    3063.623788143189,
                    3256.9325570806786,
                    3450.2413260181684,
                    3643.5500949556576,
                    3836.8588638931474,
                    4030.1676328306366,
                    4223.476401768126,
                    4416.785170705616
                ],
                "counts": [
                    2,
                    2,
                    3,
                    3,
                    1,
                    0,
                    2,
                    0,
                    4,
                    2
                ]
            },
            "statistics": [
                {
                    "name": "minimum",
                    "value": 2483.69748133072
                },
                {
                    "name": "maximum",
                    "value": 4416.785170705616
                },
                {
                    "name": "mean",
                    "value": 3416.208825291101
                },
                {
                    "name": "stddev",
                    "value": 625.4167423197122
                },
                {
                    "name": "median",
                    "value": 3172.166395824659
                }
            ],
            "sample": [
                3125.2486124401057,
                4191.843702179761,
                3172.166395824659,
                4292.653824302981,
                2774.5497654772385,
                4191.843702179761,
                4211.49568580822,
                4292.653824302981,
                4211.49568580822,
                2738.5009681369356,
                3079.6952909532847,
                4031.1221885146197,
                4211.49568580822,
                2774.5497654772385,
                3125.2486124401057,
                3172.166395824659,
                2483.69748133072,
                2738.5009681369356,
                3418.655066013459,
                4191.843702179761
            ]
        }
    }
]
```

**Comment by connormanning** [comment]
It's certainly more important to determine the content than the layout, but regardless a few notes on this initial layout:
- `statistics` makes more sense as an object with the well-known keys `minimum`/`maximum`/etc., there is no notion of an order for these keys so an array doesn't seem to fit
- `histogram` bins/counts should be combined rather than separate arrays, e.g. something along the lines of `histogram: { bin: 1234, count: 42 }`
- No need for the extra level of nesting for the `value` for each dimension, `histogram`/`statistics`/etc. can be pulled up to the top level

Regarding content, I'd rather defer to the people who have specific ideas for how they would use such information.  My proposal with #19 was more because I know that other people have uses for these things rather than a strong need of my own.  But a few thoughts and questions:
- I'm not sure `sample` provides much value, especially with mean/min/max/stddev there
- Given this VLR exists in a file, would these stats be required for *every* dimension?  What about bit flags?  What about extra bytes dimensions?
- Within a single dimension, are all of these metrics required?  Can `median` be omitted for example, as it's more difficult to compute than others?  Can `histogram` be omitted, or the entire `statistics` block?
- Can these values be estimated?  For example, can I compute them over only a certain number of octree depths corresponding to some limited resolution?  If so, how do I indicate that my values are estimates, and not full population calculations?

I don't necessarily need answers to all these since I assume the structure will evolve, but they are things to think about for anyone that wants to make modifications to this proposal.

**Comment by CCInc** [comment]
tldr; I disagree with extents by itself not being very useful. In every general use case I have, the min/max is the bare minimum of what would be useful, and the potential for a file to not have it would be devastating. I'd be interested to hear a general use-case in which having optional, more advanced stats info is superior to a required min/max.

**Comment by CCInc** [comment]
@connormanning  `For what other dimension fields would you consider min/max extents to be very useful, aside from GpsTime?`

Well, Intensity does have a fixed range within the spec, but in practice many sensors only set a small portion of that range. Given that intensity is such an environmental dependent variable, a processor would have to take the intensity the sensor gave and would have to choose to scale it to take up the entire range. Never in my experience will a sensor directly output a min of 0 and max of 65535. So, min/max would be really useful for scaling the visualizer correctly.

I personally don't think XYZ should be duplicated from the header, same with returns since they are already there. 

Classification min/max would be useful firstly to see if the file is classified without reading the entire thing in (i.e. testing if the min/max is 0-1), and also providing some expectation for whether the file falls within the ASPRS classifications (max of ~20) or it has many other classes (range of 100 or higher). 

Userdata min/max would again be useful for scaling in the visualizer, often times userdata can contain a continuous range.

Scanangle does have a discrete range but it again would be useful for scaling in the visualizer when the file doesn't utilize that entire range.

GPS-Time discussed previously

RGB would be useful to see if the file actually has RGB (i.e. the colors aren't all black), though for most uses I think this wouldn't really matter.

So I don't necessarily think the usefulness of extents would be limited to just gps-time, although for some fields it is less meaningful than others I agree.

> "Someone might like it" isn't sufficient.

Well, to me, most of the consumers of copc files want to do one thing - visualize. So to me, it's less of "someone might like it" and more of "most people will need it without knowing it".

**Comment by abellgithub** [comment]
@CCInc : You didn't really explain GPStime. Why do you want GPStime min/max, exactly?

**Comment by hobu** [comment]
#50 implemented as discussed.
